### PR TITLE
feat/assets-controller

### PR DESCRIPTION
**Pull request by leozera14**
- [x] Create SysAssetsController
- [x] Create EvmAssetsController
- [x] Create AssetsManager to export both Assets Controller and export it from MainController
- [x] Implement new Assets Controllers methods inside Import Tokens components
- [x] Create new method at MainController to update account assets based in isBitcoinBased state value
- [x] Refactor MainController methods that is using old getLatestUpdate... methods from sysweb3

### CODE REVIEW

File: source/pages/Tokens/CustomToken.tsx [hunk 1]
```diff
@@ -104,85 +45,41 @@ export const CustomToken = () => {
...
   }) => {
     setIsLoading(true);
-    setActiveNetwork(activeNetwork);
 
-    const contractResponse = (await contractChecker(
-      contractAddress,
-      activeNetwork.url
-    )) as ISupportsInterfaceProps;
+    try {
+      const provider = new ethers.providers.JsonRpcProvider(activeNetwork.url);
```
Comment (lucaslovato) [inline source/pages/Tokens/CustomToken.tsx:50]: @claudiovb is this correct? I don't know if we should call in the feature the ETH manager over the KeyringManager

File: source/pages/Tokens/CustomToken.tsx [hunk 2]
```diff
@@ -104,85 +45,41 @@ export const CustomToken = () => {
...
+        return;
+      }
+
+      await controller.wallet.account.eth.saveTokenInfo(
+        addTokenMethodResponse.tokenToAdd
+      );
+
+      setIsLoading(false);
+      setAdded(true);
+    } catch (error) {
```
Comment (lucaslovato) [inline source/pages/Tokens/CustomToken.tsx:77]: instead of calling `setIsLoading(false);` you can add it on the `finally` after the `catch` statement

File: source/pages/Tokens/ImportToken.tsx [hunk 3]
```diff
@@ -83,27 +79,29 @@ export const ImportToken: FC = () => {
...
+      await controller.wallet.assets.evm.addEvmDefaultToken(
+        token,
+        activeAccount.address,
+        activeNetwork.url
       );
 
-      const balance = `${metadata.balance / 10 ** Number(token.decimals)}`;
-      const formattedBalance = lodash.floor(parseFloat(balance), 4);
+    if (
+      typeof addTokenMethodResponse === 'boolean' ||
```
Comment (lucaslovato) [inline source/pages/Tokens/ImportToken.tsx:92]: Maybe changing this to something like `isBoolean(addTokenMethodResponse) || isNil(addTokenMethodResponse)` it might look a little bit better (both fn are from lodash)

File: source/pages/Tokens/SyscoinImport.tsx [hunk 4]
```diff
@@ -25,26 +25,27 @@ export const SyscoinImportToken = () => {
...
+    setIsLoading(true);
+
+    const addTokenMethodResponse =
+      await controller.wallet.assets.sys.addSysDefaultToken(
+        assetGuid,
+        activeNetwork.url
+      );
+
+    if (
+      typeof addTokenMethodResponse === 'boolean' ||
```
Comment (lucaslovato) [inline source/pages/Tokens/SyscoinImport.tsx:37]: same comment as above. Also in those cases why did you remove the try-catch? Maybe it is cleaner to handle the error.

File: source/scripts/Background/controllers/MainController.ts [hunk 5]
```diff
@@ -452,21 +451,82 @@ const MainController = (): IMainController => {
...
-          accountId: findAccount.id,
-          newAccountsAssets,
+        setUpdatedAllErcTokensBalance({
+          updatedTokens: newAccountsAssets,
         })
       );
     }
   };
 
+  const updateAccountAssetsValues = async () => {
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/MainController.ts:461]: 'updateAccountAssetsValues' is assigned a value but is never used

File: source/scripts/Background/controllers/account/syscoin.ts [hunk 6]
```diff
@@ -185,21 +183,26 @@ const SysAccountController = (): ISysAccountController => {
...
-      const asset = {
+      const assetInfos = {
         ...token,
         description,
-        image,
+        image: '',
         balance: Number(token.balance) / 10 ** Number(token.decimals),
       };
 
+      if (ipfsUrl !== '') {
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/account/syscoin.ts:193]: suggestion: lodash `!isEmpty(ipfsUrl)`

File: source/scripts/Background/controllers/assets/evm.ts [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+      case 'ERC-721':
+        try {
+          return await handleERC721NFTs(
+            walletAddres,
+            contractAddress,
+            symbol,
+            decimals,
+            provider
+          );
+        } catch (_erc721Error) {
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/assets/evm.ts:174]: `_` stands for non used constants

File: source/scripts/Background/controllers/assets/evm.ts [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+      // has not been founded
+      default:
+        try {
+          return await handleERC20Tokens(
+            walletAddres,
+            contractAddress,
+            decimals,
+            provider
+          );
+        } catch (_erc20Error) {
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/assets/evm.ts:193]: `_` stands for non used constants

File: source/scripts/Background/controllers/assets/evm.ts [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+          return _erc20Error;
+        }
+    }
+  };
+
+  const updateAllEvmTokens = async (
+    account: IKeyringAccountState,
+    networks: any
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/assets/evm.ts:201]: can we type `networks`?

File: source/scripts/Background/controllers/assets/syscoin.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        networkUrl,
+        xpub,
+        requestOptions,
+        true
+      );
+
+      //Validate to know which tokens use, for some cases the request only return tokens without tokensAsset
+      //and for some other cases return both
+      const validateTokenAssets =
+        tokensAsset && tokensAsset.length > 0 ? tokensAsset : tokens;
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/assets/syscoin.ts:44]: Maybe something like the example below sounds better
```ts
const isTokenAssetsValid = tokensAsset && tokensAsset.length > 0;
const validTokens = isTokenAssetsValid ? tokensAsset : tokens;
```

File: source/scripts/Background/controllers/assets/syscoin.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+      const filteredAssetsLength: ITokensAssetReponse[] = validateTokenAssets
+        ? validateTokenAssets.slice(0, 30)
+        : [];
+
+      return filteredAssetsLength;
+    } catch (error) {
+      console.log('assets by xpub error', error);
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/assets/syscoin.ts:52]: shouldn't this be a `console.error`?

File: source/scripts/Background/controllers/assets/types.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  ) => Promise<IAddCustomTokenResponse>;
+  addEvmDefaultToken: (
+    token: ITokenEthProps,
+    accountAddress: string,
+    networkUrl: string
+  ) => Promise<ITokenEthProps | boolean>;
+  updateAllEvmTokens: (
+    account: IKeyringAccountState,
+    networks: any
+  ) => Promise<any>;
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/assets/types.ts:56]: Suggestion: type networks and also the Promise return

File: source/scripts/Background/controllers/transactions/evm.ts [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+    const {
+      label: networkLabel,
+      url: networkUrl,
+      chainId: networkChainId,
+      default: isNetworkDefault,
+      apiUrl: networkApiUrl,
+    } = activeNetwork;
+
+    const etherscanSupportedNetworks = [
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/transactions/evm.ts:31]: there's no need to have that here. Maybe something like a utils/constants file to get this array
```typescript
//utils.ts or constants.ts
const ETHER_SCRAN_SUPPORTED_NETWORKS = [
   'homestead',
   'ropsten',
   'rinkeby',
   'goerli',
   'kovan',
      ];
```

File: source/scripts/Background/controllers/transactions/evm.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+      const isValidTx =
+        transaction && (from === userAddress || to === userAddress);
+
+      if (isValidTx) {
+        const { timestamp } = await wssProvider.getBlock(Number(blockNumber));
+
+        const txAlreadyExists =
+          userTransactions.ethereum.findIndex(
+            (transaction: ITransactionResponse) => transaction.hash === hash
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/transactions/evm.ts:120]: 'transaction' is already declared in the upper scope on line 108 column 13 (this might generate scope errors)
suggestion: replace one of the `transaction` const names

File: source/scripts/Background/controllers/transactions/syscoin.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+const SysTransactionController = () => {};
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/transactions/syscoin.ts:1]: `Unexpected empty arrow function`

File: source/scripts/Background/controllers/transactions/types.ts [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+  confirmations: number;
+  contractAddress: string;
+  cumulativeGasUsed: BigNumber;
+  effectiveGasPrice: BigNumber;
+  from: string;
+  gasUsed: BigNumber;
+  logs: Array<ILog>;
+  logsBloom: string;
+  transactionIndex: number;
+  status?: number;
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/transactions/types.ts:64]: Expected interface keys to be in ascending order. 'status' should be before 'transactionIndex'

File: source/scripts/Background/controllers/transactions/types.ts [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+  to: string;
+  transactionHash: string;
+  type: number;
+  root?: string;
```
Comment (lucaslovato) [inline source/scripts/Background/controllers/transactions/types.ts:68]: Expected interface keys to be in ascending order. 'root' should be before 'type'

File: source/state/vault/index.ts [hunk 18]
```diff
@@ -170,6 +170,7 @@ const VaultState = createSlice({
         value: number | string | boolean | any[];
       }>
     ) {
+      //Later with new sysweb3 change this to only get activeAccount
```
Comment (lucaslovato) [inline source/state/vault/index.ts:173]:
```suggestion
      //todo: Later with new sysweb3 change this to only get activeAccount
```

File: source/utils/tokens.ts [hunk 19]
```diff
@@ -15,10 +15,11 @@ export const getSymbolByChain = async (chain: string) => {
   return data.symbol.toString().toUpperCase();
 };
 
-export const getBalanceUpdatedToErcTokens = async (accountId: number) => {
-  const { accounts, networks } = store.getState().vault;
+//Remove this function later, will be only in EvmAssetsController
```
Comment (lucaslovato) [inline source/utils/tokens.ts:18]:
```suggestion
//todo: remove this function later will be only in EvmAssetsController
```

**Comment by lucaslovato** [comment]
@leozera14 just an FYI, tx should've been another PR maybe pointing to `feat/assets-controller` since now it is enormous and hard to review.

File: source/state/vault/types.ts [hunk 20]
```diff
@@ -11,22 +11,25 @@ export interface IVaultState {
...
-    [INetworkType.Syscoin]: {
-      [chainId: number]: INetwork;
-    };
-  };
+  networks: INetworksVault;
   timer: number;
 }
 
+export interface INetworksVault {
+  [INetworkType.Ethereum]: {
```
Comment (claudiovb) [inline source/state/vault/types.ts:25]: use from sysweb3

**Review by claudiovb — APPROVED** [review]
